### ISSUE TITLE
[Aio] Async IO manager deadlock with current API

### ISSUE THREAD

**Issue by lidizheng** [grpc/grpc#19955]
Current gRPC Python API will deadlock using Async IO manager. This problem needs to be solved before merging the first PR.

Reproduction script:
```Python
import grpc
import asyncio
from grpc.experimental import aio

async def main():
	print('===Testing Aio Channel===')
	async with aio.insecure_channel('localhost:50051') as channel:
		print(await channel.unary_unary(b'')(b''))
	print('===Finish Aio Channel===')


	print('===Testing Normal Channel===')
	with grpc.insecure_channel('localhost:50051') as channel:
		print(channel.unary_unary(b'')(b''))  # <- deadlock
	print('===Finish Normal Channel===')


aio.init_grpc_aio()
asyncio.get_event_loop().run_until_complete(main())
```

Log:
```
I0814 16:23:13.264381376  220688 call.cc:1927]               grpc_call_start_batch(call=0x564af3f8b4a0, ops=0x564af3f69a50, nops=6, tag=0x7f0903e0dc50, reserved=(nil))
I0814 16:23:13.264387822  220688 call.cc:1526]               ops[0]: SEND_INITIAL_METADATA(nil)
I0814 16:23:13.264392291  220688 call.cc:1526]               ops[1]: SEND_MESSAGE ptr=0x564af3dcbc20
I0814 16:23:13.264396790  220688 call.cc:1526]               ops[2]: SEND_CLOSE_FROM_CLIENT
I0814 16:23:13.264401459  220688 call.cc:1526]               ops[3]: RECV_INITIAL_METADATA ptr=0x7f0903e1ee88
I0814 16:23:13.264406097  220688 call.cc:1526]               ops[4]: RECV_MESSAGE ptr=0x7f0903e58f20
I0814 16:23:13.264411516  220688 call.cc:1526]               ops[5]: RECV_STATUS_ON_CLIENT metadata=0x7f0905300bf8 status=0x7f0905300c10 details=0x7f0905300c18
D0814 16:23:13.264423396  220688 dns_resolver.cc:242]        Start resolving.
I0814 16:23:13.264452976  220688 completion_queue.cc:963]    grpc_completion_queue_next(cq=0x564af3f17fd0, deadline=gpr_timespec { tv_sec: 1565824993, tv_nsec: 464450963, clock_type: 1 }, reserved=(nil))
I0814 16:23:13.464982891  220688 completion_queue.cc:963]    grpc_completion_queue_next(cq=0x564af3f17fd0, deadline=gpr_timespec { tv_sec: 1565824993, tv_nsec: 664976288, clock_type: 1 }, reserved=(nil))
I0814 16:23:13.665982102  220688 completion_queue.cc:963]    grpc_completion_queue_next(cq=0x564af3f17fd0, deadline=gpr_timespec { tv_sec: 1565824993, tv_nsec: 865976002, clock_type: 1 }, reserved=(nil))
I0814 16:23:13.866982076  220688 completion_queue.cc:963]    grpc_completion_queue_next(cq=0x564af3f17fd0, deadline=gpr_timespec { tv_sec: 1565824994, tv_nsec: 66976175, clock_type: 1 }, reserved=(nil))
```

**Comment by lidizheng** [comment]
Turned out if the main thread is blocking wait on `cq_next`, the asyncio manager will deadlock due to unable to execute another coroutine.

**Comment by lidizheng** [comment]
As discussed with @gnossen , potentially there are 4 solutions:
1. Change C-Core to support binding IO manager to each channel;
2. Switch current API to use callback stack (so both of them is using `asyncio`);
3. Introducing an open source event manager (like libuv based);
4. Wrap async API to simulate current API.

It seems the first solution is the least effort and the most clean one.

**Comment by pfreixes** [comment]
Regarding the options,

1 - Having the feeling - maybe I'm wrong - that solving this issue by adding the ability to run channels with different `iomgr` instances is not gonna be trivial, or at least the changes in the core won't be negligible. But True, that it would simulate perfectly the mixture of an async API with the sync one, where the sync one uses a different network stack which blocks the asynchronous one. Is the core team aligned with this? do they have in mind using this future feature in other use cases?

2- You mean by replacing the usage of the cq_next for the pure callback completion queue? If so, having the feeling that you would end up by doing what we were saying in the beginning, having the synchronous API build on top of the asynchronous one, having as the main problem the incapacity of keeping compatibility with Python 2.

3- IMO this would be the easiest one, if and only if this open source event manager already exists and is battle-tested. This would mean zero changes in the core, and for the `Aio` package would mean throwing away the current custom iomgr. True that we would be forced to run all callbacks triggered by the callback completion queue by using threadsafe operations which might have some performance impact, I would still consider this as a viable solution. Also, consider that implementing a POC based on that solution should be relatively easy to do.

4 - I dont get it.

**Comment by lidizheng** [comment]
For option 1, this is on the road map of C-Core team. It is possible that they will provide SWE to do it. I agree it is non-trivial, but it will benefits all wrapper languages. Particularly, the migration cases.

The `RuntimeError` solution should serve as a fallback plan. For now, I agree it is better to throw an exception, after all, exception with a reason is much nicer than deadlock.

**Comment by pfreixes** [comment]
@lidizheng could you create an issue for the fallback plan?

**Comment by lidizheng** [comment]
@pfreixes Filed #19985

**Comment by pfreixes** [comment]
Does this need to be added as part of the gRPC Async board?

**Comment by lidizheng** [comment]
After a closer look, it looks like the application thread is deadlock calling `asyncio_pollset_vtable.poll` which currently is an empty function (`pass`). The flow goes like Python -> Cython -> `cq.next()` -> C-Core -> Cython (pollset_poll) -> Python (asyncio).

However, the execution permission is never handed back to the event loop. The shortest path here to fix it is find out a way to make the `cq.next()` yield something if it is in the event loop. To be investigated.

**Comment by pfreixes** [comment]
We could explore still the solution of running the IO operations in another thread but still using an Asyncio loop. This theoretically would allow us to block the current thread while the one that holds all IO operations remains unblocked.

So, in the end, we will be having two different Asyncio loops, one for the IO operations and another one where callbacks would be executed.

The main down-side effect could be the performance, but this should be double-tested since the way both loops would have cooperation will be with a simple local pipe [1]

Another option, based on the same idea, would offload all IO operations that belong to client/server synchronous implementation to this isolated IO thread. If this could be done, only IO operations belonging to synchronous operations would be penalized. Moreover, replicate the semantic of `run_once` could be easier to implement and less disruptive with the other loop. But in any case, I do not see right now a way for identifying the context of the IO operations, so figuring out if they come from the asynchronous or the synchronous one.

WDYT on trying to make a POC of the first option? It should be something similar to the work that you already did for testing the cooperation between the EV and Asyncio, but without requiring a busy loop.

[1] https://github.com/python/cpython/blob/master/Lib/asyncio/base_events.py#L793

**Comment by pfreixes** [comment]
I've been working with a proof of concept [1] for running the IO operations in its own Asyncio loop and thread. Seems that it works, the synchronous client is still working even when the Aio module has been installed [2].

For that proof of concept I didn't bother about the `asyncio_run_loop` which semantically would need to do return the control only when a gRPC event has happened or the timeout has been reached, leveraging on the fact that internally the `cq_next` iteration would iterate calling the `asyncio_run_loop` as many times as needed till the timeout is reached. So the function ` `asyncio_run_loop` is still doing nothing with the hope that eventually the Asyncio IO loop will do its job for processing the IO operations.

I've made a performance comparison between running a pure async client against a CPP server for testing the likely footprint of having this dedicated IO loop. The results are promising, there is no noticeable impact.

The bad news is that I've detected a regression of the client performance comparing it with the original POC [3] once it gets compared with the current master. Something has been introduced in the current Aio code that slowed down the performance making it almost twice slower. This would need to be investigated.

In any case, this POC tells us that this is a doable effort. I'm wondering if we could do the same using the Event Source IO manager. Theoretically, correct me if I'm missing something @lidizheng, the same implementation could be done but having the Event Source running the separated IO loop, not needing any busy loop and having compatibility with all of the OS systems.

WDYT? /cc @gnossen @lidizheng 

[1] https://github.com/Skyscanner/grpc/commit/996636ac853acd2ce60afbb8843181321a9915b4
[2] https://gist.github.com/pfreixes/08086d17fe8357b1b75cee5aac20d561
[3] https://github.com/Skyscanner/grpc-asyncio

**Comment by lidizheng** [comment]
The regression noted. After the cancellation mechanism is done, I will look into automating the performance benchmark.

Well, I think Skyscanner@996636a is a great solution that it allows old stack and new stack to co-exist and doesn't have bad performance impact. You should make it a PR.

About using the same pattern against Event Manager, it depends on what logic we put into the callback. If the callback contains any Python object, it would be kind of messy to run in a C spawned thread. In previous experiment, it is possible to manually grab the lock and release after work. It takes time to debug all corner cases.

For the progress of Event Manager, @guantaol should be able to comment. To me, I think what you have proposed is a good enough solution to make new stack compatible with old stack, the integration of Event Manager can be a bonus item. The AsyncIO custom IO manager should be able to work with and without Event Manager.

**Comment by pfreixes** [comment]
> The regression noted. After the cancellation mechanism is done, I will look into automating the performance benchmark.

IMO we should try to find what was the root cause of this regression and fix it if its possible. Do you agree? If so, could we add an issue to the gRPC board for tracking this issue? are you planning on working on fixing this regression?

> Well, I think Skyscanner/grpc@996636a is a great solution that it allows old stack and new stack to co-exist and doesn't have bad performance impact. You should make it a PR.

Ok, Ill do that once the interceptors PR is done.

**Comment by lidizheng** [comment]
@pfreixes Yes, I think fixing the regression is a must, and we also need the mechanism to prevent further regression. So, it is not my top priority right now until we have monitoring (I'm on it, one of the block is credentials https://github.com/grpc/grpc/pull/21351).

Can you help me create the issue? I found I can't provide concrete data in the issue description, you might have more context than I do.

**Comment by pfreixes** [comment]
Added https://github.com/grpc/grpc/issues/21594, I would prefer to have this issue fixed before making any change like the one proposed for isolating the IO operations to an isolated loop, so if after the interceptors PR this issue is not yet fixed I will work on this first.

**Comment by lidizheng** [comment]
@pfreixes Thanks ahead for taking this task 👍

**Comment by pfreixes** [comment]
I'm planning to start coding the solution for supporting the synchronous stack on top of the `Aio` package Today, for circumventing the "deadlock" issue. The solution would be basically inspired by the POC that I made for having the io loop running in a separated thread [1].

Just for having all on the same page, the solution proposed would allow us to: 
- Run synchronous client requests on an already existing an initialized `Aio` environment.

The solution wouldn't provide support for: 
- Run the `Aio` package on an already initialized synchronous environment.

I'm still considering the implications of having a synchronous server running on top of an already initialized `Aio` environment, and most important if it does make sense. Initially, I would be tempted to not trying to give support to this use case. 

Also, I'm considering on removing the need of having to initialized explicitly the `Aio` package by calling the `grpc_init_aio` function, so doing this behind the scenes when either an Aio channel or an Aio server is created. For unsupported combinations, an `Exception` would be raised. This theoretically would help also to close these issues [2] [3]

/cc @gnossen @lidizheng 

[1] hthttps://github.com/grpc/grpc/issues/20806tps://github.com/Skyscanner/grpc/commit/996636ac853acd2ce60afbb8843181321a9915b4
[2]
[3] https://github.com/grpc/grpc/issues/19985

**Comment by lidizheng** [comment]
@pfreixes I think implicit `grpc_init_aio` is a good idea that prevents the application from hanging. As you suggested, we should support former case first.

**Comment by gnossen** [comment]
@pfreixes +1 to the implicit `grpc_init_aio()` call as long as we have very clear error messages when a failure occurs as a result of this call.

> I'm still considering the implications of having a synchronous server running on top of an already initialized Aio environment, and most important if it does make sense. Initially, I would be tempted to not trying to give support to this use case.

I don't think it particularly matters how we accomplish it in terms of implementation. The use case that we're enabling here is coexisting synchronous gRPC and asyncio gRPC in the same process. People need this in order to incrementally migrate from their existing sync stack to asyncio.

As for testing, it would be great if we could figure out a way to run all of the existing synchronous tests with something async added into the test process. Then we can be 100% sure that things are compatible. You might consider adding a new test runner for this.

**Comment by pfreixes** [comment]
good idea @gnossen, reuse the current sync tests definitely would be the way to go.

I guess that in the same idea of the Gevent test suite, where most likely we would need to disable the forking tests since for now we are not supporting forking in `Aio`.

**Comment by lidizheng** [comment]
This bug should be fixed by #22258.
